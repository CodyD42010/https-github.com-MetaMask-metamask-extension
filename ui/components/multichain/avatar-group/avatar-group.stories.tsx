import React from 'react';
import { StoryFn, Meta } from '@storybook/react';
import { AvatarGroup } from '.';

export default {
  title: 'Components/Multichain/AvatarGroup',
  component: AvatarGroup,
  argTypes: {
    limit: {
      control: 'number',
    },
    members: {
      control: 'object',
    },
  },
  args: {
    members: [
      { symbol: 'ETH', image: './images/eth_logo.png' },
      { symbol: 'MATIC', image: './images/matic-token.png' },
      { symbol: 'OP', image: './images/optimism.svg' },
      { symbol: 'AVAX', image: './images/avax-token.png' },
    ],
    limit: 4,
  },
} as Meta<typeof AvatarGroup>;

const Template = (args) => <AvatarGroup {...args} />;

export const DefaultStory = Template.bind({});
DefaultStory.storyName = 'Default';

export const WithTag: StoryFn<typeof AvatarGroup> = (args) => (
  <AvatarGroup {...args} />
);
WithTag.args = {
  members: [
    { symbol: 'ETH', image: './images/eth_logo.png' },
    { symbol: 'MATIC', image: './images/matic-token.png' },
    { symbol: 'OP', image: './images/optimism.svg' },
    { symbol: 'AVAX', image: './images/avax-token.png' },
    { symbol: 'PALM', image: './images/palm.svg' },
  ],
  limit: 2,
};

export const TokenWithOutSrc: StoryFn<typeof AvatarGroup> = (args) => (
  <AvatarGroup {...args} />
);
TokenWithOutSrc.args = {
  members: [
    { symbol: 'ETH', image: '' },
    { symbol: 'MATIC', image: '' },
    { symbol: 'OP', image: '' },
    { symbol: 'AVAX', image: '' },
  ],
  limit: 2,
};
