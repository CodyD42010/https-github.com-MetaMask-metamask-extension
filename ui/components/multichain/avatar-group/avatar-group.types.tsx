import { BorderColor } from '../../../helpers/constants/design-system';
import { AvatarTokenSize } from '../../component-library';
import type { StyleUtilityProps } from '../../component-library/box';

export interface AvatarGroupProps extends StyleUtilityProps {
  /** * Additional class name for the ImportTokenLink component. */
  className?: string;
  /** * Limit to show only a certain number of tokens and extras in Text */
  limit: number;
  /** * List of Avatar Tokens */
  members: {
    label: string;
    src?: string;
    size?: string;
  }[];
  /** * Size of Avatar Tokens */
  size?: AvatarTokenSize;
  /** * Border Color of Avatar Tokens */
  borderColor?: BorderColor;
}
