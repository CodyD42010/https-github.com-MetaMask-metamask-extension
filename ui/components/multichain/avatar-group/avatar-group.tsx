import * as React from 'react';
import classnames from 'classnames';
import {
  Text,
  Box,
  AvatarToken,
  AvatarTokenSize,
} from '../../component-library';
import {
  AlignItems,
  BorderColor,
  BorderRadius,
  Display,
  TextColor,
  TextVariant,
} from '../../../helpers/constants/design-system';
import { AvatarGroupProps } from './avatar-group.types';

export const AvatarGroup: React.FC<AvatarGroupProps> = ({
  className = '',
  limit = 4,
  members = [],
  size = AvatarTokenSize.Xs,
  borderColor = BorderColor.transparent,
}): JSX.Element => {
  const membersCount = members.length;
  const visibleMembers = members.slice(0, limit).reverse();
  const showTag = membersCount > limit;
  let marginLeftValue = '';
  if (AvatarTokenSize.Xs) {
    marginLeftValue = '-8px';
  } else if (AvatarTokenSize.Sm) {
    marginLeftValue = '-12px';
  } else {
    marginLeftValue = '-16px';
  }
  const tagValue = `+${(membersCount - limit).toLocaleString()}`;
  return (
    <Box
      alignItems={AlignItems.center}
      display={Display.Flex}
      className={classnames('multichain-avatar-group', className)}
      data-testid="avatar-group"
    >
      <Box display={Display.Flex}>
        {visibleMembers.map((member, i) => (
          <Box
            borderRadius={BorderRadius.full}
            key={member.symbol}
            style={
              i === 0 ? { marginLeft: '0' } : { marginLeft: marginLeftValue }
            }
          >
            <AvatarToken
              src={member.image}
              name={member.symbol}
              size={size}
              borderColor={borderColor}
            />
          </Box>
        ))}
      </Box>
      {showTag ? (
        <Box>
          <Text variant={TextVariant.bodySm} color={TextColor.textAlternative}>
            {tagValue}
          </Text>
        </Box>
      ) : null}
    </Box>
  );
};
