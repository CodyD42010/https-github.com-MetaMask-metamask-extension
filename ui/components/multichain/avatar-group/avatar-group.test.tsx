/* eslint-disable jest/require-top-level-describe */
import { render, screen } from '@testing-library/react';
import React from 'react';

import { AvatarGroup } from './avatar-group';

const members = [
  { symbol: 'ETH', image: './images/eth_logo.png' },
  { symbol: 'MATIC', image: './images/matic-token.png' },
  { symbol: 'OP', image: './images/optimism.svg' },
  { symbol: 'AVAX', image: './images/avax-token.png' },
  { symbol: 'PALM', image: './images/palm.svg' },
];

describe('AvatarGroup', () => {
  it('should render AvatarGroup component', () => {
    const { getByTestId, container } = render(
      <AvatarGroup members={members} limit={4} />,
    );
    expect(getByTestId('avatar-group')).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it('should render the tag +1 if members has a length greater than limit', () => {
    render(<AvatarGroup members={members} limit={4} />);

    expect(screen.getByText('+1')).toBeDefined();
  });

  it('should not render the tag if members has a length less than or equal to limit', () => {
    const { queryByText } = render(<AvatarGroup members={members} limit={5} />);
    expect(queryByText('+1')).not.toBeInTheDocument();
  });
});
