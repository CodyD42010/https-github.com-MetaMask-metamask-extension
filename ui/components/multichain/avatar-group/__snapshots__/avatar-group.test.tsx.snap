// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup should render AvatarGroup component 1`] = `
<div>
  <div
    class="mm-box multichain-avatar-group mm-box--display-flex mm-box--align-items-center"
    data-testid="avatar-group"
  >
    <div
      class="mm-box mm-box--display-flex"
    >
      <div
        class="mm-box 0-nidhi mm-box--rounded-full"
        style="margin-left: 0px;"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-sm mm-avatar-token mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
        >
          <img
            alt="AVAX logo"
            class="mm-avatar-token__token-image"
            src="./images/avax-token.png"
          />
        </div>
      </div>
      <div
        class="mm-box 1-nidhi mm-box--rounded-full"
        style="margin-left: -8px;"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-sm mm-avatar-token mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
        >
          <img
            alt="OP logo"
            class="mm-avatar-token__token-image"
            src="./images/optimism.svg"
          />
        </div>
      </div>
      <div
        class="mm-box 2-nidhi mm-box--rounded-full"
        style="margin-left: -8px;"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-sm mm-avatar-token mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
        >
          <img
            alt="MATIC logo"
            class="mm-avatar-token__token-image"
            src="./images/matic-token.png"
          />
        </div>
      </div>
      <div
        class="mm-box 3-nidhi mm-box--rounded-full"
        style="margin-left: -8px;"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-sm mm-avatar-token mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
        >
          <img
            alt="ETH logo"
            class="mm-avatar-token__token-image"
            src="./images/eth_logo.png"
          />
        </div>
      </div>
    </div>
    <div
      class="mm-box"
    >
      <p
        class="mm-box mm-text mm-text--body-md mm-box--color-text-default"
      >
        +1
      </p>
    </div>
  </div>
</div>
`;
